migen/build/lattice/icestorm.py

```diff
@@ -82,14 +102,39 @@ def _run_icestorm(build_name, source, yosys_opt, pnr_opt,
 
 
 class LatticeIceStormToolchain:
+    attr_translate = {
+        "keep": ("keep", "true"),
+        "no_retiming": None,  # yosys can do retiming via the (non-default)
+                              # "-retime" option to "synth_ice40", but
+                              # yosys does not check for an equivalent
+                              # constraint to prevent retiming on signals.
+        "async_reg": None,  # yosys has no equivalent, and arachne-pnr
+                            # wouldn't take advantage of it anyway.
+
+        # While custom attributes are supported in yosys, neither
+        # arachne-pnr nor icetime currently can take advantage of them
+        # to add fine-grained timing constraints.
+        "mr_ff": None,  # user-defined attribute
+        "mr_false_path": None,  # user-defined attribute
+        "ars_ff1": None,  # user-defined attribute
+        "ars_ff2": None,  # user-defined attribute
+        "ars_false_path": None,  # user-defined attribute
+
+        # ice40 does not have a shift register primitive.
+        "no_shreg_extract": None
+    }
+
     def __init__(self):
         self.yosys_opt = "-q"
+        self.pre_synthesis_commands = list()
+        self.synthesis_commands = list()
+        self.post_synthesis_commands = list()
```

Comment:
I know that this is done similarly in the other platforms and toolchains. But it's ugly and not flexible.
Can't you just have a (single) template owned by the toolchain do the job of `_build_yosys()`. You could then change the template at will instead of having `{post,pre,pnr,synth,mid...}_commands` and `{x,y,z}_opts` inflation.